### ISSUE TITLE
how inference batch data at serving?

### ISSUE THREAD

**Issue by Pekary** [OpenNMT/OpenNMT-tf#463]
I am trying serving my model using batch size greater than 1, but it seems the model take the batch data as one example. blow is my code:

```python
with graph.as_default():
    with tf.Session() as sess:
        meta_graph_def = tf.saved_model.loader.load(sess, [tf.saved_model.tag_constants.SERVING], 
                                                    export_dir = "xxx_serving")
        signature_def = meta_graph_def.signature_def["serving_default"]
        
        
        
        input0_feature = signature_def.inputs["tokens_0"].name
        input0_length = signature_def.inputs["length_0"].name

        input1_feature = signature_def.inputs["tokens_1"].name
        input1_length = signature_def.inputs["length_1"].name

        label_feature = signature_def.inputs["tokens"].name
        label_length = signature_def.inputs["length"].name

        log_probs = signature_def.outputs["log_probs"].name

        inputs_array = pad_sequences([
                ["s", "e", "q", "▁", "u", "e", "n"],
                ["s", "e", "q", "▁", "u", "e", "n", "c", "e"]], value="", dtype=object, padding="post")
        
        labels_array = pad_sequences([
                ["sequen", "ence"],
                ["sequen", "ence", "test"],], value="", dtype=object, padding="post")

        inputs = {
            input0_feature: inputs_array,
            input0_length: [9]*2,
            input1_feature: inputs_array,
            input1_length: [9]*2,
            label_feature: labels_array,
            label_length: [3]*2
        }


        log_probs = sess.run([log_probs], feed_dict=inputs)

        for log_prob in log_probs:
            print(log_prob[0])
```
no matter how many example i input, only output one log_probs list .

so i guess when inference and score operation process batch data one by one?

i know i can use loop run examples one by one, but it may be inefficiency.

so my questions is how inference batch data at serving?

**Comment by guillaumekln** [comment]
What kind of model is that?

This example is running batch inference: http://forum.opennmt.net/t/opennmt-inference-system/2760/2?u=guillaumekln

**Comment by Pekary** [comment]
the model is dual source transformer.

i find the reason, the log_probs is two dimensional array, and the first dimension is one, so the right way to do that is iterate log_probs[0].

thank you very much~

**Comment by Pekary** [comment]
hi, the padding length influences the results?  i got different results when score `[["s", "e", "q", "▁", "u", "e", "n", "", ""], ["s", "e", "q", "▁", "u", "e", "n", "c", "e"]]` and `[["s", "e", "q", "▁", "u", "e", "n"]]` for the input `["s", "e", "q", "▁", "u", "e", "n"]`, i have set the length correctly. 

but i found the onmt-main score correctly for above two inputs, it got same score for same input.

**Comment by guillaumekln** [comment]
Is it the same code as above? If yes, it does not seem the length vectors as set correctly. It should be [7, 9] instead?

**Comment by Pekary** [comment]
not the same, i set the length as [7, 9] actually.

**Comment by Pekary** [comment]
the fully code:

```python 
import tensorflow as tf
from tensorflow.contrib.seq2seq.python.ops import beam_search_ops
from keras.preprocessing.sequence import pad_sequences
import argparse


def serving(export_dir, sequence, lables=None, score_flag=False):
    if score_flag:
        assert lables is not None, "SCORE mode require labels is not None!"
    
    mode = "scores" if score_flag else "serving_default"

    graph = tf.Graph()    
    with graph.as_default():
        with tf.Session() as sess:
            meta_graph_def = tf.saved_model.loader.load(sess, [tf.saved_model.tag_constants.SERVING], 
                                                        export_dir=export_dir)
            print("meta_graph_def.signature_def keys: ")
            for k in meta_graph_def.signature_def.keys():
                print(k)
            signature_def = meta_graph_def.signature_def[mode]
            
            print("inputs keys:")
            for k in signature_def.inputs.keys():
                print(k)

            print("outputs keys:")
            for k in signature_def.outputs.keys():
                print(k)
            
            
            input0_tokens = signature_def.inputs["tokens_0"].name
            input0_length = signature_def.inputs["length_0"].name

            input1_tokens = signature_def.inputs["tokens_1"].name
            input1_length = signature_def.inputs["length_1"].name

            label_tokens = signature_def.inputs["tokens"].name
            label_length = signature_def.inputs["length"].name

            
            inputs = {
                input0_tokens: sequence["tokens"],
                input0_length: sequence["length"],
                input1_tokens: sequence["tokens"],
                input1_length: sequence["length"],
            }

            outputs = []
            if score_flag:
                outputs.append(signature_def.outputs["output"].name)

                inputs[label_tokens] = labels["tokens"]
                inputs[label_length] = labels["length"]
            else:
                outputs.append(signature_def.outputs["tokens"].name)
                outputs.append(signature_def.outputs["length"].name)

            results = sess.run(outputs, feed_dict=inputs)
            # print(results[1].shape)
            for item in zip(*results):
                if score_flag:
                    print(item[0])
                else:
                    tokens, length = item
                    length = length[0] - 1
                    print(tokens[0][:length])


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Serving for scoring and infering")
    parser.add_argument("--score", default=False, action="store_true", help="set SCORE mode, default is INFER mode")
    parser.add_argument("--export_dir", type=str, default="my_export_dir", help="Model export directory")
    args = parser.parse_args()
    inputs_array = pad_sequences([
                    ["s", "e", "q", "▁", "u", "e", "n"],
                    ["s", "e", "q", "▁", "u", "e", "n", "c", "e"]], value="", dtype=object, padding="post")
    inputs_length = [7, 9]

    sequence = {
        "tokens": inputs_array,
        "length": inputs_length
    }

    labels = None
    if args.score:
        labels_array = pad_sequences([
                ["sequen", "ence"],
                ["sequen", "ence", "test"],], value="", dtype=object, padding="post")

        labels_length = [2, 3]
        labels = {
            "tokens": labels_array,
            "length": labels_length
        }

    serving(args.export_dir, sequence, labels, args.score)
```

**Comment by guillaumekln** [comment]
Are you serving a custom model?

**Comment by Pekary** [comment]
yes, dual source transformer, and i need export graph for scoring.

**Comment by guillaumekln** [comment]
How did you define the scoring graph?

**Comment by Pekary** [comment]
I change the code at [estimator.py](https://github.com/OpenNMT/OpenNMT-tf/blob/master/opennmt/estimator.py#L214) to:
```python
elif mode == tf.estimator.ModeKeys.PREDICT:
      labels["alignment"] = None
      outputs, predictions = local_model(features, labels, params, mode)

      # Forward example index for reordering predictions.
      if "index" in features:
        predictions["index"] = features["index"]
      
      cross_entropy = tf.nn.sparse_softmax_cross_entropy_with_logits(
        logits=outputs["logits"], labels=labels["ids_out"])
      
      weights = tf.sequence_mask(labels["length"], dtype=cross_entropy.dtype)
      masked_cross_entropy = cross_entropy * weights
      scores = tf.reduce_sum(masked_cross_entropy, axis=1, name="scores")

      export_outputs = {}
      export_outputs[tf.saved_model.signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY] = (
          tf.estimator.export.PredictOutput(predictions))
      export_outputs[tf.saved_model.signature_constants.CLASSIFY_OUTPUT_SCORES] = (
          tf.estimator.export.PredictOutput(scores))

      return tf.estimator.EstimatorSpec(
          mode,
          predictions=predictions,
          export_outputs=export_outputs)

    else:
      raise ValueError("Invalid mode")
```

**Comment by guillaumekln** [comment]
As far as I understand, there is no issue with OpenNMT-tf so I will not be able to spend too much time helping you. You should probably debug your score graph for any missing/incorrect masking.

**Comment by Pekary** [comment]
ok, thanks~

**Comment by Pekary** [comment]
i found the problem, it maybe that [your examples](http://forum.opennmt.net/t/opennmt-inference-system/2760/2?u=guillaumekln) also have the problem. the problem is blow:
```python
inputs = {
        input_tokens: [
            ["Hello", "world", "!", ""],
            ["How", "are", "you", "?"]],
        input_length: [3, 4]
    }
```
if using inputs tokens as above, we will get 
```python
features["ids_out"] = [
        ["Hello", "world", "!", "", "</s>"],
        ["How", "are", "you", "?", "</s>"]
    ]
```
but we need this:
```python
features["ids_out"] = [
        ["Hello", "world", "!", "</s>", ""],
        ["How", "are", "you", "?", "</s>"]
    ]
```
